airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py

```diff
@@ -0,0 +1,349 @@
+        for property_name in properties:
+            this_property_name = property_name
+            if property_name.startswith("$"):
+                # Just remove leading '$' for 'reserved' mixpanel properties name, example:
+                # from API: '$browser'
+                # to stream: 'browser'
+                this_property_name = this_property_name[1:]
+            record[this_property_name] = properties[property_name]
+
+        return record
+
+
+class RevenueDpathExtractor(DpathExtractor):
+    def extract_records(self, response: requests.Response) -> List[Mapping[str, Any]]:
+        """
+        response.json() example:
+        {
+            'computed_at': '2021-07-03T12:43:48.889421+00:00',
+            'results': {
+                '$overall': {       <-- should be skipped
+                    'amount': 0.0,
+                    'count': 124,
+                    'paid_count': 0
+                },
+                '2021-06-01': {
+                    'amount': 0.0,
+                    'count': 124,
+                    'paid_count': 0
+                },
+                '2021-06-02': {
+                    'amount': 0.0,
+                    'count': 124,
+                    'paid_count': 0
+                },
+                ...
+            },
+            'session_id': '162...',
+            'status': 'ok'
+        }
+        """
+        new_records = []
+        for record in super().extract_records(response):
+            for date_entry in record:
+                if date_entry != "$overall":
+                    list.append(new_records, {"date": date_entry, **record[date_entry]})
+        return new_records
+
+
+class FunnelsDpathExtractor(DpathExtractor):
+    def extract_records(self, response: requests.Response) -> List[Mapping[str, Any]]:
+        """
+        response.json() example:
+        {
+            'computed_at': '2021-07-03T12:43:48.889421+00:00',
+            'results': {
+                '$overall': {       <-- should be skipped
+                    'amount': 0.0,
+                    'count': 124,
+                    'paid_count': 0
+                },
+                '2021-06-01': {
+                    'amount': 0.0,
+                    'count': 124,
+                    'paid_count': 0
+                },
+                ...
+            },
+            'session_id': '162...',
+            'status': 'ok'
+        }
+        """
+        new_records = []
+        for record in super().extract_records(response):
+            for date_entry in record:
+                list.append(new_records, {"date": date_entry, **record[date_entry]})
+        return new_records
+
+
+class FunnelsSubstreamPartitionRouter(SubstreamPartitionRouter):
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        """
+        Add 'funnel_name' to the slice, the rest code is exactly the same as in super().stream_slices(...)
+        """
+        if not self.parent_stream_configs:
```

Comment:
Is it possible here to loop over `super().stream_slices(...)` and add the `funnel_name` field into `StreamSlice` instead of rewriting whole method?

Reply:
there is no source of funnel_name. It can be taken from parent stream only in this method.

Reply:
can't we use a normal substream partition router and an AddField component to add the extra field?

Reply:
1. The most important thing is: we need to update stream_slices here because this is the only place where we can populate more data (funnel_name field) from parent stream 
2. we inject funnel_name using AddField component from stream_slice, because if we add funnel_name  to partition then it will appear in stream state and we will not be able to make proper Lecacy state migration

Reply:
I'm not sure I follow.

It sounds like this could be a generic substream partition router if the stream slice followed the usual pattern of `"parent_slice": parent_partition` instead of `"funnel_name": parent_record.get("name")`.

It's not clear to me why that's not possible.

> make proper Lecacy state migration
I would generally prefer custom state migration over custom partition routing

Reply:
From the parent record we should take 2 fields: 'funnel_id' and 'name'

'id' field is populated by default "substream partition router" component

but in order to populate 'name' parent_record we need to add this custom component